### PR TITLE
[feat] update multiselect filter and add input dropdown

### PR DESCRIPTION
**Pull request by jinkang-0**
[//]: # "These comments are meant for your reference. They are invisible and don't need to be deleted!"

### :palm_tree: What's new in this PR
[//]: # "REQUIRED - Describe what's new in this PR in a few lines. A description and bullet points for specifics will suffice."

Added components:
- `DropdownMenu`
    - `DropdownMenu.Item`
- `InputDropdown`
- `DropdownMenu`

Dropdown Details
- Refactored dropdown meu to have visibility toggled with `display: none` instead of unmounting it.
- Removed dropdown's dependency on `pSBC`, opens up room to remove it in the future
- Added dropdown open animation
- Added input dropdown caret rotation animation
- Added keyboard-friendly navigation (tab, arrow keys, enter)

Other Improvements
- Modified `app/cases/page.tsx` to use filters from HIFI

### :evergreen_tree: Screenshots
[//]: # "REQUIRED for frontend changes, otherwise optional but strongly recommended. Add screenshots of expected behavior - GIFs if you're feeling fancy! If you are making changes here, please CC: @davidqing6432 at the bottom."

<details>
<summary>Filter dropdown - updated styles</summary>
<br />

Dropdown-ception. 🪆

demo_filter
</details>

<details>
<summary>Filter dropdown - keyboard navigation</summary>
<br />

Keys:
- Tab and shift tab to navigate between filter buttons
    - Tabbing when the menu is open will close it
- Up and down arrow keys to navigate between dropdown items
- Enter (in menu) to select/unselect dropdown items
- Enter (menu closed, no item focused) to expand/collapse dropdown menu
- Escape to close the menu

demo_filter_keyboard

</details>

<details>
<summary>Input dropdown - basic functionality</summary>
<br />

demo_basic
</details>

<details>
<summary>Input dropdown - keyboard navigation</summary>
<br />

Keys:
- Tab or shift tab to navigate between inputs
- Up and down arrow keys to navigate between dropdown items
- Enter (when an item is selected) to select/deselect an item
- Backspace (when input is empty) to delete last added item

demo_keyboard

</details>

<details>
<summary>Input dropdown - handling overflow</summary>
<br />
Demo of handling selected tag overflow and text input overflow

demo_overflow
</details>

<details>
<summary>Input dropdown - error and disabled</summary>
<br />
Sorry, no gif here, it got disabled and errored.

image
</details>

## :deciduous_tree: How to review
[//]: # 'REQUIRED - Describe the order in which to review files and what to expect when testing locally. Is there anything specifically you want feedback on? Should this be reviewed commit by commit, or all at once? What are some user flows to test? What are some edge cases to look out for?'

Standard procedure:
```
git fetch origin jinkang/ijp-55-update-multiselect-filter
git checkout jinkang/ijp-55-update-multiselect-filter
npm run dev
```

if you've fetched before, make sure to update it:
```
git checkout jinkang/ijp-55-update-multiselect-filter
git pull origin jinkang/ijp-55-update-multiselect-filter
```

To review `InputDropdown`:
- Setup the testing page with the example code in `InputDropdown/index.tsx`

Recommended order:
- First review the cases page for new filter dropdowns
- Then make a new page using the example usage (commented at the bottom of `InputDropdown`)

or be a rebel and go in any order.

Either way, try to break it in any way you can.

## :seedling: Next steps 
[//]: # "Optional - What's NOT in this PR, doesn't work yet, and/or still needs to be done. Note any temporary fixes in this PR that should be cleaned up later."

Consult David on whether or not we need a separate dropdown item design for the selected + hovered state.

### :information_source: Online sources
[//]: # 'Optional - copy links to any tutorials or documentation that was useful to you when working on this PR'

### :potted_plant: Related PRs
[//]: # "Optional - related PRs you're waiting on/ PRs that will conflict, etc; if this is a refactor, feel free to add PRs that previously modified this code"

[//]: # 'This tags the project leader as a default. Feel free to change, or add on anyone who you should be in on the conversation.'
CC: @pragyakallanagoudar, @davidqing6432

### CODE REVIEW

File: src/app/cases/page.tsx [hunk 1]
```diff
@@ -48,64 +57,67 @@ export default function Page() {
       <H1>Browse Available Cases</H1>
       <FiltersContainer>
         <FilterDropdown
-          defaultValue="Remote/In Person"
-          options={['Remote/In Person', 'Remote Only', 'In Person Only']}
-          onChange={v =>
-            setCaseFilters({ ...caseFilters, remote: v as string })
-          }
+          defaultValue={defaultFilterValues.remote}
```
Comment (pragyakallanagoudar) [inline src/app/cases/page.tsx]: question for @davidqing6432: personally, I'm not a huge fan of the fact that both options being selected shows up as `Remote + 1 more` or `In Person + 1 more` depending on which one was selected first...

some alternatives I can think of:
- handle this one filter differently by making it so that if both options are selected, it just says `Remote/In Person` (don't love this because selecting everything has the same effect as selecting nothing, but looks different because the filter is filled in/blue)
- make the filter single-select?? so you'd only filter if you want to restrict options to one of the two?

File: src/app/cases/page.tsx [hunk 2]
```diff
@@ -48,64 +57,67 @@ export default function Page() {
...
+          multi
+          options={new Set(['Remote', 'In Person'])}
+          value={caseFilters.remote}
+          onChange={v => setCaseFilters({ ...caseFilters, remote: v })}
         />
         <FilterDropdown
```
Comment (pragyakallanagoudar) [inline src/app/cases/page.tsx:78]: same feelings on this filter as the one above

File: src/app/cases/page.tsx [hunk 3]
```diff
@@ -48,64 +57,67 @@ export default function Page() {
...
         />
       </FiltersContainer>
       <MainDisplay>
         <CardColumn>
           {caseData
             .filter(c => {
-              if (caseFilters.remote === 'remote') return c.is_remote;
-              if (caseFilters.remote === 'inperson') return !c.is_remote;
-              return true;
+              if (caseFilters.remote.size === 0) return true;
```
Comment (pragyakallanagoudar) [inline src/app/cases/page.tsx]: relating to the other comment on this filter: it just feels a bit strange to me that `caseFilters.remote.size === 0` and `caseFilters.remote.size === 2` are essentially the same 🤔

File: src/app/cases/page.tsx [hunk 4]
```diff
@@ -48,64 +57,67 @@ export default function Page() {
...
+              if (caseFilters.remote.has('Remote') && c.is_remote) return true;
+              if (caseFilters.remote.has('In Person') && !c.is_remote)
+                return true;
```
Comment (pragyakallanagoudar) [inline src/app/cases/page.tsx]: nit: can we condense this somehow? maybe

```
c => caseFilters.remote.size === 0 || (caseFilters.remote.has('Remote') && c.is_remote) || (caseFilters.remote.has('In Person') && !c.is_remote)
```

File: src/app/cases/page.tsx [hunk 5]
```diff
@@ -48,64 +57,67 @@ export default function Page() {
...
-                  .filter(c => c.country)
-                  .map(c => (c.country ? c.country : '')),
-              ),
-            ),
-          ]}
-          onChange={v =>
-            setCaseFilters({ ...caseFilters, countries: v as string[] })
+          options={
+            new Set(
+              caseData.filter(c => c.country).map(c => c.country),
```
Comment (pragyakallanagoudar) [inline src/app/cases/page.tsx:105]: can we only pull cases that need either an attorney or an interpreter? i think some of the cases that are initially displayed are not seeking either role, which we don't want to display at all.

File: src/components/DropdownMenu/styles.tsx [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+    to {
+      opacity: 1;
+      z-index: 999;
+      transform: translateY(0);
+    }
+  }
+`;
+
+// menu option
+export const DropdownItem = styled.p<{
```
Comment (pragyakallanagoudar) [inline src/components/DropdownMenu/styles.tsx:50]: the checkboxes look FANTASTIC!!! thank you for your work on this

File: src/components/InputDropdown/index.tsx [hunk 7]
```diff
@@ -0,0 +1,353 @@
...
+              'Raspberry',
+              'Strawberry',
+              'Tomato',
+              'Vanilla Bean',
+              'Watermelon',
+              'Zucchini',
+            ])
+          }
+        />
+        <InputDropdown
```
Comment (pragyakallanagoudar) [inline src/components/InputDropdown/index.tsx:339]: question: is this dropdown meant to constantly demonstrate an error state? i.e. since it always seems to be displaying the error message

File: src/components/InputDropdown/index.tsx [hunk 8]
```diff
@@ -0,0 +1,353 @@
...
+              'Pencil',
+              'Mechanical Pencil',
+              'Chalk',
+              'Etch A Sketch',
+              'Quill',
+              'Mouse',
+            ])
+          }
+        />
+        <InputDropdown
```
Comment (pragyakallanagoudar) [inline src/components/InputDropdown/index.tsx:357]: CLEANNNNN

File: src/components/DropdownMenu/styles.tsx [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+  background: white;
+  box-shadow: 0 2px 0.25em 0.1em rgba(0, 0, 0, 0.25);
+  border: 1px solid rgba(0, 0, 0, 0.25);
+  padding: 0.2rem 0.3rem;
+  border-radius: 0.5rem;
+  z-index: 100;
+
+  &:empty {
+    &::after {
+      content: 'No matches found!';
```
Comment (pragyakallanagoudar) [inline src/components/DropdownMenu/styles.tsx]: two things:
- nit: can we get rid of the exclamation point in the no matches found message
- when i type something in and then delete it completely, it seems to go back to this "no matches found" state. is this the expected behavior? i think it might be better to display every option in the dropdown, since the search string is empty

File: src/components/InputDropdown/index.tsx [hunk 10]
```diff
@@ -0,0 +1,353 @@
...
+
+    document.addEventListener('click', globalClickEvent);
+
+    return () => {
+      document.removeEventListener('click', globalClickEvent);
+    };
+  }, []);
+
+  // handle when a tag gets clicked
+  function handleRemoveOption(option: string) {
```
Comment (pragyakallanagoudar) [inline src/components/InputDropdown/index.tsx]: is there some way to make it so that the tag is not deleted when it is clicked, but rather, when the 'X' on the tag is clicked?

File: src/components/InputDropdown/index.tsx [hunk 11]
```diff
@@ -0,0 +1,353 @@
...
+
+  // decide add/remove options
+  function handleSelectOption(option: string) {
+    if (inputRef.current) inputRef.current.innerHTML = '';
+
+    if (multi) {
+      const copy = new Set(value);
+
+      if (value.has(option)) copy.delete(option);
+      else copy.add(option);
```
Comment (pragyakallanagoudar) [inline src/components/InputDropdown/index.tsx:97]: not sure if this is where we would handle this, but: when the tags overflow to the next line, it seems like the cursor moves to the line after:

Screenshot 2023-11-18 at 6 17 58 PM

is there some way to change this so the cursor stays on the same line?

File: src/components/InputDropdown/index.tsx [hunk 12]
```diff
@@ -0,0 +1,353 @@
...
+              'Watermelon',
+              'Zucchini',
+            ])
+          }
+        />
+        <InputDropdown
+          label="Writing Utensil"
+          placeholder="Pen"
+          value={writingUtensil}
+          setValue={setWritingUtensil}
```
Comment (pragyakallanagoudar) [inline src/components/InputDropdown/index.tsx:343]: question for @davidqing6432:

thoughts on having the single-select dropdown have checkboxes too? lowk i feel like that might be more consistent, but idk what the design convention is

File: src/components/InputDropdown/index.tsx (hunk 7 above)
Comment (pragyakallanagoudar) [reply src/components/InputDropdown/index.tsx:339]: and if so, just to clarify, we would want to eventually use a `useState` for the error message right

**Review by pragyakallanagoudar — COMMENTED** [review]
LOOKS AMAZING JINKANGGGG thank you so much for your hard work on this :DDDD

have listed some comments/questions above, if there's anything else you would like feedback on that I missed please let me know

File: src/app/cases/page.tsx (hunk 3 above)
Comment (jinkang-0) [reply src/app/cases/page.tsx]: I agree, but I think we had this discussion before and I thought the consensus reached was that multiselect all the way was better?

File: src/app/cases/page.tsx (hunk 4 above)
Comment (jinkang-0) [reply src/app/cases/page.tsx]: ye, thanks for flagging - forgot about these filters tbh

File: src/app/cases/page.tsx (hunk 5 above)
Comment (jinkang-0) [reply src/app/cases/page.tsx:105]: Do we want to update the query for that or get rid of them entirely from the database?

File: src/components/InputDropdown/index.tsx (hunk 7 above)
Comment (jinkang-0) [reply src/components/InputDropdown/index.tsx:339]: ye, if there's no error, we can leave it as `''`

File: src/components/DropdownMenu/styles.tsx (hunk 9 above)
Comment (jinkang-0) [reply src/components/DropdownMenu/styles.tsx]: yes 🫡 thanks for the bug catch

File: src/components/InputDropdown/index.tsx (hunk 10 above)
Comment (jinkang-0) [reply src/components/InputDropdown/index.tsx]: I think this is more of a question for @davidqing6432 
I personally prefer making it delete when the tag is clicked so users don't have to pinpoint the exact 'X', but I also understand it could be odd that the whole thing disappears when you click on it

also for @davidqing6432 (while you're here) thoughts on adding hover and active effects to the tag? (maybe stretch)

File: src/components/InputDropdown/index.tsx (hunk 11 above)
Comment (jinkang-0) [reply src/components/InputDropdown/index.tsx:97]: i'll try

File: src/app/cases/page.tsx (hunk 5 above)
Comment (pragyakallanagoudar) [reply src/app/cases/page.tsx:105]: let's update the query!

File: src/app/cases/page.tsx (hunk 3 above)
Comment (pragyakallanagoudar) [reply src/app/cases/page.tsx]: yep you're right! i think my worry with the single select filter having one option be plural (i.e. `In Person/Remote`) still stands. but yes this also doesn't feel like an ideal solution, so i definitely see the motivations for originally having that sort of filter..let's have a discussion about this during next worksession! cc @davidqing6432

File: src/app/cases/page.tsx (hunk 1 above)
Comment (jinkang-0) [reply src/app/cases/page.tsx]: another alternative is to make it so it's sorted (i.e. always `Remote +1 more` when both are chosen, regardless of the order). thoughts?

File: src/components/DropdownMenu/styles.tsx (hunk 9 above)
Comment (jinkang-0) [reply src/components/DropdownMenu/styles.tsx]: as for the second one, I don't think I'm experiencing this? After deleting the query, it displays all the options for me.

Comment (jinkang-0) [reply src/components/DropdownMenu/styles.tsx]: Ah, I think I know what you're talking about. If you type in a query to search, then select an option, your query will be deleted but the filter will remain. This is intended. Two reasons:
1. The user can keep typing to search without having to manually delete their existing query
2. The user may want to select more options in their existing query

Comment (pragyakallanagoudar) [reply src/components/DropdownMenu/styles.tsx]: hmm i think it might be different? i.e. i typed a bunch of gibberish without selecting any options, the menu said "no matches found", and then i backspaced completely and it still said "no matches found"
Screenshot 2023-11-19 at 3 57 25 PM

File: src/components/InputDropdown/index.tsx (hunk 10 above)
Comment (pragyakallanagoudar) [reply src/components/InputDropdown/index.tsx]: agree! can see it going both ways

File: src/components/InputDropdown/index.tsx [hunk 13]
```diff
@@ -12,9 +12,11 @@ import {
   DropdownErrorText,
   DropdownInput,
   DropdownInputContainer,
+  DropdownInputFrame,
```
Comment (pragyakallanagoudar) [inline src/components/InputDropdown/index.tsx:16]: agh just pulled and tested and for some reason it seems like this issue is still present for me....(i.e. dropdown clickable outside of the box)

Comment (pragyakallanagoudar) [reply src/components/InputDropdown/index.tsx:16]: should probably get everyone to test during tue worksesh, i will add to agenda

File: src/app/cases/page.tsx (hunk 1 above)
Comment (pragyakallanagoudar) [reply src/app/cases/page.tsx]: definitely think the sorted approach would also improve things! we should prob discuss all of this at worksesh (it's on the agenda)

File: src/components/DropdownMenu/styles.tsx (hunk 9 above)
Comment (jinkang-0) [reply src/components/DropdownMenu/styles.tsx]: I found a way to reproduce the screenshot you sent by pasting some random code with trailing new lines. I implemented a fix to trim white spaces, which seems to fix it. Let me know if the issue persists!

File: src/api/supabase/queries/cases.ts [hunk 14]
```diff
@@ -2,13 +2,20 @@ import { UUID } from 'crypto';
...
 export async function getNCases(n: number) {
-  const { data, error } = await supabase.from('cases').select().limit(n);
+  const { data, error } = await supabase
+    .from('cases')
+    .select()
+    .or('needs_interpreter.eq.true,needs_attorney.eq.true')
+    .limit(n);
   if (error) throw new Error(`Error fetching ${n} cases: ${error.message}`);
   return data;
 }
```
Comment (varortz) [inline src/api/supabase/queries/cases.ts:21]: we might need to update these after the switch to using the database function for the queries. May or may not also need to update the database func?

File: src/api/supabase/queries/cases.ts [hunk 15]
```diff
@@ -2,13 +2,20 @@ import { UUID } from 'crypto';
 import supabase from '../createClient';
 
 export async function getAllCases() {
```
Comment (varortz) [inline src/api/supabase/queries/cases.ts:4]: we might need to update these after the switch to using the database function for the queries. May or may not also need to update the database func?

File: src/api/supabase/queries/cases.ts [hunk 16]
```diff
@@ -2,13 +2,20 @@ import { UUID } from 'crypto';
...
-  const { data, error } = await supabase.from('cases').select();
+  const { data, error } = await supabase
+    .from('cases')
+    .select()
+    .or('needs_interpreter.eq.true,needs_attorney.eq.true');
```
Comment (varortz) [inline src/api/supabase/queries/cases.ts:8]: we might need to update these after the switch to using the database function for the queries. May or may not also need to update the database func?

Comment (jinkang-0) [reply src/api/supabase/queries/cases.ts:8]: yeah we should update the database function to check that as well! I can update the code here so hopefully there's no merge conflicts

File: src/components/DropdownMenu/styles.tsx (hunk 9 above)
Comment (pragyakallanagoudar) [reply src/components/DropdownMenu/styles.tsx]: ahdjhjdewd it seems to still be there :( let's also test for this at worksesh tmr!!

File: src/app/cases/page.tsx (hunk 1 above)
Comment (ccatherinetan) [reply src/app/cases/page.tsx]: i agree with the sorted approach!

**Comment by jinkang-0** [comment]
closing this PR for a new implementation to use `react-select`
